src/sql/workbench/contrib/welcome/page/browser/welcomePage.ts

```diff
@@ -0,0 +1,1066 @@
+import { registerColor, focusBorder, descriptionForeground, foreground, contrastBorder, activeContrastBorder } from 'vs/platform/theme/common/colorRegistry';
+import { getExtraColor } from 'vs/workbench/contrib/welcome/walkThrough/common/walkThroughUtils';
+import { IExtensionsWorkbenchService } from 'vs/workbench/contrib/extensions/common/extensions';
+import { IEditorInputFactory, EditorInput } from 'vs/workbench/common/editor';
+import { INotificationService, Severity } from 'vs/platform/notification/common/notification';
+import { TimeoutTimer } from 'vs/base/common/async';
+import { areSameExtensions } from 'vs/platform/extensionManagement/common/extensionManagementUtil';
+import { ILabelService } from 'vs/platform/label/common/label';
+import { IFileService } from 'vs/platform/files/common/files';
+import { ExtensionType } from 'vs/platform/extensions/common/extensions';
+import { joinPath } from 'vs/base/common/resources';
+import { IRecentlyOpened, isRecentWorkspace, IRecentWorkspace, IRecentFolder, isRecentFolder, IWorkspacesService } from 'vs/platform/workspaces/common/workspaces';
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { IHostService } from 'vs/workbench/services/host/browser/host';
+import { IProductService } from 'vs/platform/product/common/productService';
+
+
+const configurationKey = 'workbench.startupEditor';
+const oldConfigurationKey = 'workbench.welcome.enabled';
+const telemetryFrom = 'welcomePage';
+
+export class WelcomePageContribution implements IWorkbenchContribution {
+
+	constructor(
+		@IInstantiationService instantiationService: IInstantiationService,
+		@IConfigurationService configurationService: IConfigurationService,
+		@IEditorService editorService: IEditorService,
+		@IBackupFileService backupFileService: IBackupFileService,
+		@IFileService fileService: IFileService,
+		@IWorkspaceContextService contextService: IWorkspaceContextService,
+		@ILifecycleService lifecycleService: ILifecycleService,
+		@ICommandService private readonly commandService: ICommandService,
+	) {
+		const enabled = isWelcomePageEnabled(configurationService, contextService);
+		if (enabled && lifecycleService.startupKind !== StartupKind.ReloadedWindow) {
+			backupFileService.hasBackups().then(hasBackups => {
+				const activeEditor = editorService.activeEditor;
+				if (!activeEditor && !hasBackups) {
+					const openWithReadme = configurationService.getValue(configurationKey) === 'readme';
+					if (openWithReadme) {
+						return Promise.all(contextService.getWorkspace().folders.map(folder => {
```

Comment:
Do we want this to happen in the background without waiting on it? The constructor isn't going to wait for this promise or the callback on line 68 to finish.

Reply:
This was brought over from VS Code

Reply:
Yeah but technical debt is still technical debt. We have code like this in other places but at least "new" code we're putting into ADS shouldn't contain these bugs.

Like when we use async in constructors it makes it really difficult to actually make sure everything is done.

Although I don't have context if we're ever going to need to make sure the things the constructor is doing is done before using the rest of the code.

@Charles-Gagnon what do you think?

Reply:
So yeah - in general we shouldn't be doing this kind of thing. But I'm not sure we can avoid it - at some point this is inherently going to need to run an async function to load up this data. Doing it from a constructor isn't ideal but we don't really have much of a choice here since is called by the contribution point. And in the end it's just basically kicking off a bunch of async loads that eventually call and open the actual welcome page so the fact that this returns without completing seems fine to me (although personally I'd refactor this into an async function that is called by the constructor since that let's us use await and keep things much cleaner)